File: DSA/Anisha_Parveen/Day_11_Ques_1.cpp
#include <bits/stdc++.h>
using namespace std;

int directions[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

int minMoves(vector<vector<char>> &grid, pair<int, int> start, pair<int, int> goal)
{
    int rows = grid.size();
    int cols = grid[0].size();
    vector<vector<bool>> visited(rows, vector<bool>(cols, false));
    queue<pair<pair<int, int>, int>> q;
    q.push({start, 0});
    visited[start.first][start.second] = true;

    while (!q.empty())
    {
        auto current = q.front();
        q.pop();
        int x = current.first.first;
        int y = current.first.second;
        int moves = current.second;

        if (x == goal.first && y == goal.second)
        {
            return moves;
        }

        for (int i = 0; i < 4; ++i)
        {
            int newX = x + directions[i][0];
            int newY = y + directions[i][1];

            if (newX >= 0 && newX < rows && newY >= 0 && newY < cols && grid[newX][newY] == '.' && !visited[newX][newY])
            {
                visited[newX][newY] = true;
                q.push({{newX, newY}, moves + 1});
            }
        }
    }

    return -1;
}

int main()
{
    vector<vector<char>> grid = {
        {'.', '.', '.'},
        {'.', 'X', '.'},
        {'.', '.', '.'}};
    pair<int, int> start = {0, 0};
    pair<int, int> goal = {2, 2};

    cout << minMoves(grid, start, goal) << endl;

    return 0;
}

File: DSA/Asmi_Garg/day11_ques1.java
import java.util.*;

public class Main {
    static int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    public static int minMoves(char[][] grid, int[] start, int[] goal) {
        int rows = grid.length;
        int cols = grid[0].length;
        Queue<int[]> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        queue.add(start);
        visited.add(start[0] + "," + start[1]);
        int moves = 0;

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] pos = queue.poll();
                if (pos[0] == goal[0] && pos[1] == goal[1]) {
                    return moves;
                }
                for (int[] dir : directions) {
                    int x = pos[0] + dir[0];
                    int y = pos[1] + dir[1];
                    if (x >= 0 && x < rows && y >= 0 && y < cols && grid[x][y] != 'X' && !visited.contains(x + "," + y)) {
                        queue.add(new int[]{x, y});
                        visited.add(x + "," + y);
                    }
                }
            }
            moves++;
        }

        return -1;  // return -1 if goal is not reachable
    }

    public static void main(String[] args) {
        char[][] grid = {
                {'.', '.', '.'},
                {'.', 'X', '.'},
                {'.', '.', '.'}
        };
        int[] start = {0, 0};
        int[] goal = {2, 2};

        System.out.println(minMoves(grid, start, goal));  // Output: 4
    }
}
